### PR TITLE
Commit with new futures and fixes (fixes #7)

### LINKED ISSUE #7: Add to codebase the custom assertions, data utility class for data retrieval, SPI
- Create custom assertions for Client and Panel entities;
 - Create an interface Reader for reading data from a custom data source;
 - Create concrete implementations for reading data from json and xml format;
 - Create a generic utility class for data retrieval;
 - Create several tests, which read data from different external data sources;
 - Add Service Provider Interface;

### PR DESCRIPTION
**Pull request by mkgerasimenko**
- Created custom assertions for Client and Panel entities;
 - Created an interface Reader for reading data from a custom data source;
 - Created concrete implementations for reading data from json and xml format;
 - Created a generic utility class for data retrieval;
 - Created several tests, which read data from different external data sources;
 - Added Service Provider Interface;

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -8,16 +12,40 @@ repositories {
...
+}
+
 dependencies {
-    compile('org.testng:testng:6.12')
+    compile("org.testng:testng:${testngVersion}",
+            "io.github.sskorol:test-data-supplier:${dataSupplierVersion}",
+            "org.projectlombok:lombok:${lombokVersion}",
+            "org.assertj:assertj-core:${assertjVersion}",
+            "com.google.code.gson:gson:${gsonVersion}",
+            'com.fasterxml.jackson.dataformat:jackson-dataformat-xml:2.9.2',
```
Comment (sskorol) [inline build.gradle]: do you want to move other versions to `ext` section as well?

File: build.gradle [hunk 2]
```diff
@@ -8,16 +12,40 @@ repositories {
     jcenter()
 }
 
+ext{
```
Comment (sskorol) [inline build.gradle]: seems like a missing whitespace... do you want to reformat code?

File: build.gradle [hunk 3]
```diff
@@ -8,16 +12,40 @@ repositories {
...
+            'org.jdom:jdom2:2.0.6')
 }
 
 task wrapper(type: Wrapper) {
     gradleVersion '4.2.1'
 }
 
 test {
+    testLogging {
+        showStandardStreams = false
```
Comment (sskorol) [inline build.gradle]: that's redundant if you're not using it

File: src/main/java/com/waverleysoftware/assertions/ClientAssert.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+public class ClientAssert extends AbstractAssert<ClientAssert, Client> {
+    private static final String ERROR_MESSAGE_TEMPLATE = "\nExpecting <%s> of: \n<%s>\nto be: <%s>\n but was : \n <%s>";
+
+    public ClientAssert(final Client client) {
+        super(client, ClientAssert.class);
+    }
+
+    public ClientAssert hasClientName(final String clientName) {
+        isNotNull();
+        final String name = actual.getFirstName();
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/assertions/ClientAssert.java]: split methods calling with variables assignment to increase code readability...

File: src/main/java/com/waverleysoftware/assertions/ClientAssert.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        if (!Objects.areEqual(name, clientName)) {
+            failWithMessage(ERROR_MESSAGE_TEMPLATE, "Client first name", actual, clientName, name);
+        }
+        return this;
+    }
+
+    public ClientAssert hasMasterCode(final int conditionForAge) {
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/assertions/ClientAssert.java]: how `masterCode` is related to `age`?

File: src/main/java/com/waverleysoftware/assertions/CustomAssertions.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package com.waverleysoftware.assertions;
+
+import com.waverleysoftware.model.Client;
+import com.waverleysoftware.model.Panel;
+
+/**
+ * A simple base class for Assertions.
+ */
+@SuppressWarnings("Warning")
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/assertions/CustomAssertions.java]: which warning are you suppressing?

File: src/main/java/com/waverleysoftware/listemers/SecondTestListener.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    public void onStart(final ISuite suite) {
+        log.info("Start!!!!");
+    }
+
+    @Override
+    public void onFinish(final ISuite suite) {
+        log.info("Finish!!!!");
+    }
+
+    public static Reader getCurrentImpl(final String dataSource) {
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/listemers/SecondTestListener.java]: give it more obvious name... what does "current impl" mean in your app's context?

File: src/main/java/com/waverleysoftware/listemers/SecondTestListener.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        final String extension = capitalize(dataSource.substring(dataSource.lastIndexOf('.')).replace(".", ""));
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/listemers/SecondTestListener.java]: `commons-lang3` library which you are using, contains powerful string utils class, where you can get a substring after last character occurrence

File: src/main/java/com/waverleysoftware/utils/ImplLoader.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import com.waverleysoftware.utils.api.Reader;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
+
+/**
+ * A simple service-provider loading facility.
+ */
+public final class ImplLoader {
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/utils/ImplLoader.java]: make it more generic... it should be a service loader for any interface you provide

File: src/main/java/com/waverleysoftware/listemers/SecondTestListener.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        final Optional<Reader> readerOptional = StreamEx.of(projectImplementations)
+                .filter(impl -> impl.toString().contains(extension))
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/listemers/SecondTestListener.java]: there's a method called `endsWith`... you may want to use it instead...

File: src/main/java/com/waverleysoftware/listemers/SecondTestListener.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+                .findFirst();
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/listemers/SecondTestListener.java]: you can replace previous filter with conditional `findFirst`

File: src/main/java/com/waverleysoftware/listemers/SecondTestListener.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        return readerOptional.orElseGet(readerOptional::get);
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/listemers/SecondTestListener.java]: no need to save result into variable if you return it on next line...

File: src/main/java/com/waverleysoftware/model/Client.java [hunk 13]
```diff
@@ -0,0 +1,21 @@
+package com.waverleysoftware.model;
+
+import lombok.Data;
+
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * A simple model class for panel processing.
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/model/Client.java]: it doesn't seem like a panel

File: src/main/java/com/waverleysoftware/utils/ImplLoader.java [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    private ImplLoader() {
+        throw new UnsupportedOperationException();
+    }
+
+    public static List<Reader> getAllProjectImplementation() {
+        final List<Reader> implList = new ArrayList<>();
+        final ServiceLoader<Reader> serviceLoader = ServiceLoader.load(Reader.class);
+        for (Reader reader : serviceLoader) {
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/utils/ImplLoader.java]: use streams instead

File: src/main/java/com/waverleysoftware/utils/api/Reader.java [hunk 15]
```diff
@@ -0,0 +1,8 @@
+package com.waverleysoftware.utils.api;
+
+/**
+ * A simple interface for read of file processing.
+ */
+public interface Reader {
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/utils/api/Reader.java]: give it more obvious name

File: src/test/java/com/waverleysoftware/AccessCodeTests.java [hunk 16]
```diff
@@ -1,32 +1,36 @@
...
-    public void setUp() {
-        System.out.println("Setting up on child level");
-    }
+import static java.util.Arrays.asList;
 
-    @Test(priority = 1)
-    public void accessCodeShouldBeCreated() {
-        System.out.println("Access code created");
+/**
+ * A simple model class for access code testing.
```
Comment (sskorol) [inline src/test/java/com/waverleysoftware/AccessCodeTests.java]: it doesn't seem like a model class

File: src/test/java/com/waverleysoftware/AccessCodeTests.java [hunk 17]
```diff
@@ -1,32 +1,36 @@
...
+ */
+@Slf4j
+public class AccessCodeTests extends BaseTest {
+    @Test(dataProvider = "getPanelData", dataProviderClass = DataSuppliers.class)
+    public void accessCodeShouldBeAdded(final Panel... panels) {
+        final String conditionForType = "home";
```
Comment (sskorol) [inline src/test/java/com/waverleysoftware/AccessCodeTests.java]: it makes no sense to create a separate local variable, if you then use it on next like while assertion

File: src/test/java/com/waverleysoftware/AccessCodeTests.java [hunk 18]
```diff
@@ -1,32 +1,36 @@
...
+    public String getName() {
+        return "someName";
     }
 
-    @Test(priority = 3)
-    public void accessCodeShouldBeDeleted() {
-        System.out.println("Access code deleted");
+    @Test(dataProvider = "getPanelData", dataProviderClass = DataSuppliers.class)
+    public void accessCodeShouldBeCreatedWithSomeDates(final Panel... panels) {
+        SOFTASSERT.assertThat(asList(panels).get(0).getName()).as("panel name").isEqualTo("lyn");
```
Comment (sskorol) [inline src/test/java/com/waverleysoftware/AccessCodeTests.java]: if you're always using first element, why don't you return just a single object from data supplier?

File: src/test/java/com/waverleysoftware/BaseTest.java [hunk 19]
```diff
@@ -1,16 +1,42 @@
...
-        System.out.println("Setting up on parent level");
+    public void setUp(final ITestContext context) {
+        log.info(context.getCurrentXmlTest().getName());
+        log.info("{} - {}", "Starting process on parent level", LocalDate.now());
+    }
+
+    @BeforeMethod
+    public void setUp(final ITestContext context, final Method method) {
+        StreamEx.of(method.getDeclaredAnnotations())
+                .forEach(System.out::println);
```
Comment (sskorol) [inline src/test/java/com/waverleysoftware/BaseTest.java]: it's not recommended to use `sout` in production code... replace it with logger

File: src/test/java/com/waverleysoftware/DataFromFileTest.java [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * A simple model class for some data testing.
+ */
+@Slf4j
+public class DataFromFileTest extends BaseTest {
+    @Data(source = "panel.json")
+    @Test(dataProvider = "getPanelDataFromFile", dataProviderClass = DataSuppliers.class)
+    public void readDataFromFileJson(final Panel panel) {
+        CustomAssertions.assertThat(panel).hasPanelName("adc");
```
Comment (sskorol) [inline src/test/java/com/waverleysoftware/DataFromFileTest.java]: use static import to avoid class specification before method call

**Review by sskorol — CHANGES_REQUESTED** [review]
@mkgerasimenko check comments please

**Comment by mkgerasimenko** [comment]
@sskorol, please look at the latest changes for this branch.

File: src/test/java/com/waverleysoftware/data/DataSuppliers.java [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @DataSupplier(flatMap = true)
+    public <T> T getData(final Method method) {
+        return getTypeByProvidedInfo(method);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> T getTypeByProvidedInfo(final Method method) {
+        final Data declaredAnnotation = method.getDeclaredAnnotation(Data.class);
```
Comment (sskorol) [inline src/test/java/com/waverleysoftware/data/DataSuppliers.java]: you can put it into `ofNullable`

File: src/main/java/com/waverleysoftware/assertions/ClientAssert.java [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+
+        final String name = actual.getFirstName();
+
+        if (!Objects.areEqual(name, clientName)) {
+            failWithMessage(ERROR_MESSAGE_TEMPLATE, "Client first name", actual, clientName, name);
+        }
+        return this;
+    }
+
+    public ClientAssert hasNextAge(final int clientAge) {
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/assertions/ClientAssert.java]: so what does `next` mean in this context?

**Review by sskorol — CHANGES_REQUESTED** [review]
@mkgerasimenko check comments please

**Review by sskorol — APPROVED** [review]
LGTM

**Pull request merged**